tcp_proxy: support command operators in tunneling_config.hostname

jewertow opened the pull request:
Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

Commit Message: tcp_proxy: support command operators in tunneling_config.hostname
Additional Description: This change enables dynamically setting `tunneling_config.hostname` with command operators.
Risk Level: Low
Testing: added unit tests
Docs Changes: done
Release Notes: done
Platform Specific Features: none
[Optional Fixes #19612 #21804]

This pull request is an alternative for [auto_sni](https://github.com/envoyproxy/envoy/pull/20230).

This change allows to configure TCP proxy as follows:
```
tunneling_config:
  hostname: %REQUESTED_SERVER_NAME%:443
```

jewertow commented:
@ramaraochavali @lambdai

ramaraochavali reviewed (commented):
Thank you. LGTM

jewertow commented:
@alyssawilk could you take a look at this pull request? I didn't add integration test yet, but you can review the general idea and implementation. This approach makes it possible to support also other [command operators](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#command-operators) if make sense. The advantage of this approach in contrary to overwriting `Host` header is that the `hostname` is not empty nor set to a dummy value what would be confusing.

I gave up the idea of `auto_sni`, because it's not backward compatible to wrap `hostname` and `auto_sni` in `oneof`. So the idea does not make sense, because it would be required to define empty or dummy hostname and I wanted to avoid it.

kyessenov commented on source/common/tcp_proxy/tcp_proxy.h:
```diff
@@ -117,7 +118,16 @@ class TunnelingConfigHelperImpl : public TunnelingConfigHelper {
+    int port = std::stoi(hostname_.substr(colon_idx + 1));
+    if (downstream_info.has_value() && hostname == "%REQUESTED_SERVER_NAME%") {
+      return fmt::sprintf(
+          "%s:%d", downstream_info->get().downstreamAddressProvider().requestedServerName(), port);
```
I wonder why we don't just support the entire formatter dictionary here? It's a bit strange and error-prone to just pick one variable.

alyssawilk replied:
yeah it's great to see this hanging together end to end, but the implementation seems pretty fragile.  I'd suggest a standalone utility to swap the host part of the host-part pair, 
Also maybe a string_view for requested_server_name and a utility to handle the optional deref rather than passing all the downstream info through, WDYT?
/wait

jewertow replied:
I don't see a need to support values like: `%START_TIME%` or `%UPSTREAM_PROTOCOL%`. Why would anyone need other values from dictionary?

kyessenov replied:
@jewertow Yes, those are not applicable, but I think headers, things like metadata, filter state, or other TLS attributes (DNS SAN) are reasonable.

jewertow replied:
Unfortunately headers couldn't be used, because it's a TcpProxy, not an HttpConnectionManager, so these data are not available. When it comes to TLS attributes, there are only available SANs of downstream proxy (`%DOWNSTREAM_LOCAL_URI_SAN%`, `%DOWNSTREAM_PEER_URI_SAN%`) and using these values does not make sense, because it would connect back to the downstream - I can't imagine any real use case. Other TLS attributes don't make sense in the context of hostname.

I agree that dynamic metadata or filter state might somehow be useful to provide hostname, but only these two. I considered all other command operators and can't imagine that any of them would be useful in the context of hostname. So I don't agree that supporting all command operators is important, because almost all of them are useless and it's not worth to implement it, test all cases and then maintain useless features and complex code base.

> It's a bit strange and error-prone to just pick one variable.

I don't agree that it's error-prone, because passing a command operator which is evaluated to irrelevant value is also strange error-prone. I'm going to document it, so it won't be confusing nor strange.

I understand your motivation and respect your feedback, but I just don't believe in this idea and I really think it's useless.
Dynamic metadata and filter state might be somehow useful, but I suggest to not support them as far as users don't need them. This PR aims to solve a real and common problem, so I am not convinced to extend its scope if it's not needed.

kyessenov replied:
I am not disagreeing about the value or that it only makes sense with a subset of formatter strings. I am trying to avoid having special custom parsing code since it's difficult to maintain. I'm OK just supporting SNI for now, and tackle metadata/filter state later. As I mentioned earlier, we'll have to support URL templates for MASQUE which need generic templates.

kyessenov commented on source/common/tcp_proxy/tcp_proxy.h:
```diff
@@ -117,7 +118,15 @@ class TunnelingConfigHelperImpl : public TunnelingConfigHelper {
+  const std::string host(const absl::optional<absl::string_view> requested_server_name =
+                             absl::nullopt) const override {
+    auto [hostname, port] =
+        Network::Utility::hostAndPortFromRequestTargetInAuthorityForm(hostname_);
```
Does this throw an exception when `hostname_` has no `:`? I think we allowed hosts without full port form before.
Also, could we avoid parsing it every time? Would make more sense to only parse once and reject bad config rather than throwing on data path.

kyessenov replied:
Checking on the spec: I think CONNECT requires `:` but it's no longer required in extended CONNECT. I also didn't find any place where it's enforced, so this probably deserves better validation on configuration.

kyessenov replied:
Extended CONNECT was added later to h2 to support websockets and other non-TCP payloads: https://datatracker.ietf.org/doc/html/rfc8441.

kyessenov reviewed (commented):
Please make sure exceptions are not thrown on the data path, otherwise, looks reasonable.

KBaichoo commented:
@kyessenov I think this is waiting on your reply; PTAL

@jewertow seems like you need to merge main

lambdai commented:
Ping @jewertow 
Are you still working on it? It's an easy one to remove the exception. 

I'd like to propose the richer formatter on top of this change. Let me know if you want to complete this one

lambdai commented:
/wait-any restore the flag

howardjohn commented on test/common/tcp_proxy/upstream_test.cc:
```diff
@@ -382,6 +382,32 @@ TYPED_TEST(HttpUpstreamRequestEncoderTest, RequestEncoderHeadersWithDownstreamIn
+
+TYPED_TEST(HttpUpstreamRequestEncoderTest,
+           RequestEncoderHostnameWithDownstreamInfoRequestedServerName) {
+  this->config_message_.set_hostname("%REQUESTED_SERVER_NAME%:443");
```
Can we support arbitrary header additions? See https://github.com/envoyproxy/envoy/issues/21804. This is a major blocker for us after some recent changes landed. Given we can add arbitrary metadata for all headers _except_ `:authority` today it seems like this should be feasible?

jewertow replied:
This change aims to allow setting command operators in the `tunneling_config.hostname` field to avoid setting empty or dummy hostname and then overriding header "Host" or ":authority" which are not allowed to override. But as @kyessenov requested, I'm refactoring this implementation to rely on substitution formatter which supports `%DYNAMIC_METADATA%`, so this change should provide what you need.

jewertow commented on test/common/tcp_proxy/upstream_test.cc:
```diff
@@ -383,7 +404,55 @@ TYPED_TEST(HttpUpstreamRequestEncoderTest, RequestEncoderHeadersWithDownstreamIn
+}
+
+TYPED_TEST(HttpUpstreamRequestEncoderTest,
+           RequestEncoderHostnameWithDownstreamInfoDynamicMetadata) {
```
cc: @lambdai

howardjohn replied:
awesome, thanks!

jewertow commented:
@kyessenov I implemented this feature with substitution formatter as you requested and added a unit test for `DYNAMIC_METADATA`. Could you review once again?

kyessenov commented on source/common/tcp_proxy/tcp_proxy.h:
```diff
@@ -114,17 +114,26 @@ class TunnelingConfigHelperImpl : public TunnelingConfigHelper {
   const bool use_post_;
   std::unique_ptr<Envoy::Router::HeaderParser> header_parser_;
+  Formatter::FormatterPtr hostname_fmt_;
+  const Http::RequestHeaderMapPtr empty_request_headers_;
```
This is uninitialized. You should store by value or better use `Http::StaticEmptyHeaders`.

jewertow replied:
Fixed

kyessenov commented:
/wait

jewertow commented on envoy/tcp/upstream.h:
```diff
@@ -25,8 +25,10 @@ class GenericUpstream;
   // The host name of the tunneling upstream HTTP request.
-  virtual const std::string& hostname() const PURE;
+  // This function evaluates command operators if specified. Otherwise it returns host name as is.
+  virtual const std::string host(const StreamInfo::StreamInfo& stream_info) const PURE;
```
Done

jewertow commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -77,8 +79,15 @@ Config::SharedConfig::SharedConfig(
-    tunneling_config_helper_ =
-        std::make_unique<TunnelingConfigHelperImpl>(config.tunneling_config());
+    const auto& tunneling_config = MessageUtil::downcastAndValidate<
+        const envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy_TunnelingConfig&>(
```
Done

jewertow commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -77,8 +79,15 @@ Config::SharedConfig::SharedConfig(
+        const envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy_TunnelingConfig&>(
+        config.tunneling_config(), context.messageValidationVisitor());
+    envoy::config::core::v3::SubstitutionFormatString sfs;
+    sfs.mutable_text_format_source()->set_inline_string(tunneling_config.hostname());
```
I added comments to proto and briefly described the change in the current changelog.

jewertow commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -77,8 +79,15 @@ Config::SharedConfig::SharedConfig(
+    const auto& tunneling_config = MessageUtil::downcastAndValidate<
+        const envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy_TunnelingConfig&>(
+        config.tunneling_config(), context.messageValidationVisitor());
+    envoy::config::core::v3::SubstitutionFormatString sfs;
```
Fixed

kyessenov commented on changelogs/current.yaml:
```diff
@@ -32,6 +32,9 @@ behavior_changes:
     envoy_redis_myprefix_command_pttl_latency_sum{} to envoy_redis_command_pttl_latency_sum{envoy_redis_prefix="myprefix"}.
+- area: tcp_proxy
+  change: |
+    added support for command operators in :ref:`TunnelingConfig hostname <envoy_v3_api_field_envoy.config.filter.network.tcp_proxy.v3.TcpProxy.TunnelingConfig.hostname>` to dynamically set upstream hostname.
```
I think the label should be:
`envoy_v3_api_field_extensions.filters.network.tcp_proxy.v3.TcpProxy.TunnelingConfig.hostname`

jewertow replied:
Thanks, I fixed it.

kyessenov commented on source/common/tcp_proxy/tcp_proxy.h:
```diff
@@ -114,17 +115,25 @@ class TunnelingConfigHelperImpl : public TunnelingConfigHelper {
   const std::string hostname_;
   const bool use_post_;
   std::unique_ptr<Envoy::Router::HeaderParser> header_parser_;
+  Formatter::FormatterPtr hostname_fmt_;
```
Make this `const Formatter::FormatterPtr`.

jewertow replied:
I can't, because `fromProtoConfig` does not return `const`.

lizan replied:
You can if you initialize this in initialization list and extract
```
envoy::config::core::v3::SubstitutionFormatString substitution_format_config;
  substitution_format_config.mutable_text_format_source()->set_inline_string(
      config_message.hostname());
  return Formatter::SubstitutionFormatStringUtils::fromProtoConfig(
      substitution_format_config, context);
```
to a helper function.

kyessenov commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -77,8 +79,14 @@ Config::SharedConfig::SharedConfig(
   if (config.has_tunneling_config()) {
-    tunneling_config_helper_ =
-        std::make_unique<TunnelingConfigHelperImpl>(config.tunneling_config());
+    envoy::config::core::v3::SubstitutionFormatString substitution_format_config;
```
I think this code could be made part of `TunnelingConfigHelperImpl`. E.g. add a function there, and initialize const formatter field with a function call.

jewertow replied:
Good idea, I did it.

kyessenov commented on test/common/tcp_proxy/upstream_test.cc:
```diff
@@ -40,17 +42,25 @@ template <typename T> class HttpUpstreamTest : public testing::Test {
+    Formatter::FormatterPtr hostname_fmt =
+        Formatter::SubstitutionFormatStringUtils::fromProtoConfig(substitution_format_config,
+                                                                  context_);
+    config_ = std::make_unique<TunnelingConfigHelperImpl>(config_message_, std::move(hostname_fmt));
```
If you moved formatter construction into helper you wouldn't have to duplicate this logic several times.

jewertow replied:
Done

kyessenov reviewed (changes requested):
Generally looks good, but there are a few comments that could improve this I think.

kyessenov commented:
@jewertow Please don't force push. That makes it hard to track previous comments. Thanks.
/wait

kyessenov commented on test/common/tcp_proxy/upstream_test.cc:
```diff
@@ -207,14 +210,23 @@ template <typename T> class HttpUpstreamRequestEncoderTest : public testing::Tes
   }
 
+  void populateMetadata(envoy::config::core::v3::Metadata& metadata, const std::string ns,
+                        const std::string key, const std::string value) {
```
nit: these could be `const std::string&`

jewertow replied:
Done

kyessenov commented on source/common/tcp_proxy/tcp_proxy.h:
```diff
@@ -114,17 +116,29 @@ class TunnelingConfigHelperImpl : public TunnelingConfigHelper {
-      : hostname_(config_message.hostname()), use_post_(config_message.use_post()),
-        header_parser_(Envoy::Router::HeaderParser::configure(config_message.headers_to_add())) {}
-  const std::string& hostname() const override { return hostname_; }
+          config_message,
```
nit: It's better to move this implementation code to `.cc` file to prevent header pollution. also `this->` is unusual, but not a big deal.

jewertow replied:
Ok, I moved constructor and `host` method to .cc file.

kyessenov reviewed (approved):
LGTM, need @envoyproxy/api-shepherds approval for proto change, and a couple of code nits.

jewertow commented:
/retest

lizan reviewed (commented):
/lgtm api
